master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,667 @@
+		} else {
+			c.spotLoopState.inBackoffState = false
+		}
+	}
+}
+
+func (c *awsCluster) cleanup(ctx *actor.Context, allSpotRequests []*ec2.SpotInstanceRequest) {
+	// For requests that are in a terminal state, clean up any orphaned instances. For requests that
+	// might require user attention to fix, log the error, but make sure we only log the error once
+
+	instancesToTerminate := make([]*string, 0, 0)
+	allRequestsInApi := make(map[string]bool)
+	spotRequestsToNotifyUserAbout := make([]*ec2.SpotInstanceRequest, 0, 0)
+
+	for _, request := range allSpotRequests {
+		allRequestsInApi[*request.SpotInstanceRequestId] = true
+		switch *request.Status.Code {
+		case
+			"cancelled-before-fulfillment",
+			"instance-terminated-by-price",
+			"instance-terminated-by-schedule",
+			"instance-terminated-by-service",
+			"instance-terminated-by-user",
+			"spot-instance-terminated-by-user",
+			"instance-terminated-launch-group-constraint",
+			"instance-terminated-no-capacity",
+			"marked-for-stop",
+			"marked-for-termination",
+			"instance-stopped-by-price",
+			"instance-stopped-by-user",
+			"instance-stopped-no-capacity",
+			"schedule-expired",
+			"system-error":
+			// Unfulfillable, not requiring cleanup
+			continue
+		case "request-canceled-and-instance-running":
+			// Unfulfillable, maybe requiring cleanup
+			instancesToTerminate = append(instancesToTerminate, request.InstanceId)
+			continue
+		case
+			"bad-parameters",
+			"constraint-not-fulfillable",
+			"limit-exceeded":
+			// Unfulfillable, requiring user attention
+			if _, ok := c.spotLoopState.onlyLogErrorOnceTracker[*request.SpotInstanceRequestId]; !ok {
+				spotRequestsToNotifyUserAbout = append(spotRequestsToNotifyUserAbout, request)
+				c.spotLoopState.onlyLogErrorOnceTracker[*request.SpotInstanceRequestId] = true
+			}
+			continue
+		}
+	}
+
+	// Log errors in chronological order
+	sort.SliceStable(spotRequestsToNotifyUserAbout, func(i, j int) bool {
+		return spotRequestsToNotifyUserAbout[i].CreateTime.Before(*spotRequestsToNotifyUserAbout[j].CreateTime)
+	})
+	for _, request := range spotRequestsToNotifyUserAbout {
+		ctx.Log().
+			WithField("spot-request-status-code", request.Status.Code).
+			WithField("spot-request-status-message", request.Status.Message).
+			WithField("spot-request-creation-time", request.CreateTime).
+			Error("a spot request cannot be fulfilled and may require user intervention")
+	}
+
+	_, err := c.terminateInstances(instancesToTerminate, false)
+	if err != nil {
+		ctx.Log().WithError(err).Error("cannot terminate EC2 instances associated with inactive spot requests")
+	}
+
+	// Keep the size of the onlyLogErrorOnceTracker small by removing items after
+	// the AWS API stops keeping track of the request
+	for spotRequestId, _ := range c.spotLoopState.onlyLogErrorOnceTracker {
+		if _, ok := allRequestsInApi[spotRequestId]; !ok {
+			delete(c.spotLoopState.onlyLogErrorOnceTracker, spotRequestId)
+		}
+	}
+
+	return
+}
+
+// EC2 calls
```

Comment:
This comment should be in a separate paragraph rather than attracted to a function